File: framework/languages/dotnet/samples/Bingo/Shared/Contracts/Messages.cs
namespace Bingo.Shared.Contracts;

public sealed record AuthenticateReq(string AccessToken);

public sealed record AuthenticateRes(
    string ActorId,
    string DisplayName);

public sealed record AuthenticatePlayerReq(string AccessToken);

public sealed record AuthenticatePlayerRes(
    bool Accepted,
    string? ActorId,
    string? DisplayName,
    string? Reason);

public sealed record EnsurePlayerActorReq(
    string ActorId,
    string DisplayName);

public sealed record ActorRefSnapshot(
    byte[] NodeRid,
    string ActorId,
    ulong Generation);

public sealed record EnsurePlayerActorRes(
    string ActorId,
    string ActorType,
    ActorRefSnapshot Actor);

public sealed record MatchBingoReq(string Mode);

public sealed record MatchBingoRes(
    string RoomId,
    BingoRoomState State);

public sealed record MatchBingoApiReq(
    string ActorId,
    string DisplayName,
    string Mode);

public sealed record MatchBingoApiRes(string RoomId);

public sealed record AllocateBingoRoomReq(string Mode);

public sealed record AllocateBingoRoomRes(string RoomId);

public sealed record BingoRoomJoinReq(
    string RoomId,
    string ActorId,
    string DisplayName);

public sealed record BingoRoomJoinRes(BingoRoomState State);

public sealed record StartBingoGameReq(string RoomId);

public sealed record StartBingoGameRes(BingoRoomState State);

public sealed record LeaveRoomReq(string RoomId);

public sealed record LeaveRoomRes(BingoRoomState State);

public sealed record PlayerJoinedNotify(
    string RoomId,
    string ActorId,
    string DisplayName,
    int Seat,
    bool IsHost,
    BingoRoomState State);

public sealed record BingoGameStartedNotify(BingoRoomState State);

public sealed record BingoNumberDrawnNotify(
    string RoomId,
    int DrawSeq,
    int Number,
    BingoRoomState State);

public sealed record BingoStateNotify(BingoRoomState State);

public sealed record BingoGameEndedNotify(BingoRoomState State);

public sealed record BingoRoomState(
    string RoomId,
    string Status,
    string HostActorId,
    bool CanStart,
    int DrawSeq,
    int? LastDrawnNumber,
    IReadOnlyList<int> DrawnNumbers,
    IReadOnlyList<BingoPlayerState> Players,
    IReadOnlyList<string> Winners);

public sealed record BingoPlayerState(
    string ActorId,
    string DisplayName,
    int Seat,
    bool IsHost,
    int[] Card,
    bool[] Marks,
    int CompletedLines);

File: framework/languages/java/samples/java/Bingo/Shared/src/main/java/systems/zlink/samples/bingo/shared/contracts/Messages.java
package systems.zlink.samples.bingo.shared.contracts;

import java.util.List;
import systems.zlink.framework.handlers.ZLinkPacket;

public final class Messages {
    private Messages() {
    }

    @ZLinkPacket("AuthenticateReq")
    public record AuthenticateReq(String accessToken) {
    }

    public record AuthenticateRes(String actorId, String displayName, String actorNodeRid) {
    }

    @ZLinkPacket("AuthenticatePlayer")
    public record AuthenticatePlayerReq(String accessToken) {
    }

    public record AuthenticatePlayerRes(boolean accepted, String actorId, String displayName, String reason) {
    }

    @ZLinkPacket("EnsurePlayerActor")
    public record EnsurePlayerActorReq(String actorId, String displayName, String preferredActorNodeRid) {
    }

    public record ActorRefWire(String nodeRid, String actorId, long generation) {
    }

    public record EnsurePlayerActorRes(String actorId, String actorType, ActorRefWire actor) {
    }

    @ZLinkPacket("MatchBingoReq")
    public record MatchBingoReq(String mode) {
    }

    public record MatchBingoRes(String roomId, BingoRoomState state, String roomOwnerNodeRid) {
    }

    @ZLinkPacket("MatchBingoApiReq")
    public record MatchBingoApiReq(String actorId, String displayName, String mode, String actorNodeRid) {
    }

    public record MatchBingoApiRes(String roomId, String roomOwnerNodeRid) {
    }

    @ZLinkPacket("AllocateBingoRoomReq")
    public record AllocateBingoRoomReq(String actorId, String mode, String preferredOwnerNodeRid) {
    }

    public record AllocateBingoRoomRes(String roomId, String roomOwnerNodeRid) {
    }

    @ZLinkPacket("BingoRoomJoinReq")
    public record BingoRoomJoinReq(String roomId, String actorId, String displayName, boolean observeOnly) {
    }

    public record BingoRoomJoinRes(BingoRoomState state) {
    }

    @ZLinkPacket("SubmitBingoCardReq")
    public record SubmitBingoCardReq(String roomId, List<Integer> card) {
    }

    public record SubmitBingoCardRes(BingoRoomState state) {
    }

    public record PlayerJoinedNotify(
        String roomId,
        String actorId,
        String displayName,
        int seat,
        boolean isHost,
        BingoRoomState state) {
    }

    public record BingoGameStartedNotify(BingoRoomState state) {
    }

    public record BingoNumberDrawnNotify(
        String roomId,
        int drawSeq,
        int number,
        BingoRoomState state) {
    }

    public record BingoStateNotify(BingoRoomState state) {
    }

    public record BingoGameEndedNotify(BingoRoomState state) {
    }

    @ZLinkPacket("ObserveBingoEventsReq")
    public record ObserveBingoEventsReq(String roomId) {
    }

    public record ObserveBingoEventsRes(boolean subscribed, String observerNodeRid) {
    }

    @ZLinkPacket("StopObservingBingoEventsReq")
    public record StopObservingBingoEventsReq(String roomId) {
    }

    public record StopObservingBingoEventsRes(boolean stopped, String observerNodeRid) {
    }

    public record BingoWinnerMsg(
        String roomId,
        String actorId,
        int drawSeq,
        String itemId,
        String itemName,
        String rarity) {
    }

    public record BingoRewardAnnouncedNotify(
        String roomId,
        String actorId,
        int drawSeq,
        String itemId,
        String itemName,
        String rarity,
        String receivingSpotNodeRid) {
    }

    public record BingoWinner(String actorId) {
    }

    public record BingoRoomState(
        String roomId,
        String status,
        String hostActorId,
        boolean canStart,
        int drawSeq,
        Integer lastDrawnNumber,
        List<Integer> drawnNumbers,
        List<BingoPlayerState> players,
        List<String> winners) {
    }

    public record BingoPlayerState(
        String actorId,
        String displayName,
        int seat,
        boolean isHost,
        List<Integer> card,
        List<Boolean> marks,
        int completedLines) {
    }
}
